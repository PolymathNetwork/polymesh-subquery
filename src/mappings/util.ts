import { Codec } from "@polkadot/types/types";
import { u8aToString } from "@polkadot/util";
/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str.replace(/[A-Z]/g, (letter) => `_${letter.toLowerCase()}`);

export const capitalizeFirstLetter = (str: string): string =>
  str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s.replace(/\0/g, "");

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (
  text: string,
  exitOnFirst = false
): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case "(":
      case "{":
      case "<": {
        nestedLevel++;
        break;
      }
      case ")":
      case "}":
      case ">": {
        nestedLevel--;
        break;
      }
      case ",": {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(
      `No top level comma found in ${text}, it probably isn't a map`
    );
  }
  return commas;
};

export const getOrDefault = <K, V>(
  map: Map<K, V>,
  key: K,
  getDefault: () => V
): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};
