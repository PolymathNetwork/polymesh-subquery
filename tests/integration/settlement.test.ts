import { gql } from '@apollo/client/core';
import { getApolloClient } from '../util';
const { query } = getApolloClient();

describe('settlements', () => {
  test('with no filter', async () => {
    const q = {
      query: gql`
        query {
          settlements(first: 10) {
            totalCount
            nodes {
              blockId
              addresses
              legs
              result
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });

  test('with from filter', async () => {
    const q = {
      query: gql`
        query {
          settlements(
            filter: {
              legs: {
                contains: [
                  {
                    from: {
                      did: "0x0500000000000000000000000000000000000000000000000000000000000000"
                    }
                  }
                ]
              }
            }
          ) {
            totalCount
            nodes {
              id
              blockId
              legs
              addresses
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });

  test('with ticker filter', async () => {
    const q = {
      query: gql`
        query {
          settlements(filter: { legs: { contains: [{ ticker: "4TICKER" }] } }) {
            totalCount
            nodes {
              id
              blockId
              legs
              addresses
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });

  test('with address filter', async () => {
    const q = {
      query: gql`
        query {
          settlements(
            filter: {
              addresses: {
                contains: ["0xbc18d7b0241882bcfc94b5fe0d520988f2a9f4050a2a1df58c4de14ee23f3915"]
              }
            }
          ) {
            totalCount
            nodes {
              id
              blockId
              legs
              addresses
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });
});
