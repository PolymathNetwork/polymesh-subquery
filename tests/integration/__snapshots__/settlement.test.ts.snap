// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSettlements test getting settlements 1`] = `
Object {
  "settlements": Object {
    "__typename": "SettlementsConnection",
    "nodes": Array [
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GKLEq7o3zZCGAeokuSoAMbmxamJ2uXJ984Fg8xYpaaWMUPr",
        ],
        "blockId": 139,
        "legs": Array [
          Object {
            "amount": "100",
            "from": Object {
              "did": "0xc5e2d554233da63d509ee482dbeed0ddec94dc1d0b45ebfdcdc48bd0928222b1",
              "number": "0",
            },
            "ticker": "4TICKER",
            "to": Object {
              "did": "0xc5e2d554233da63d509ee482dbeed0ddec94dc1d0b45ebfdcdc48bd0928222b1",
              "number": "1",
            },
          },
        ],
        "result": "Executed",
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5DcLNmYXCC8J1hLHJsbtPFNX7kSDx4YTbgGkNYcHUKdvbgU9",
        ],
        "blockId": 262,
        "legs": Array [
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "11ATICKER",
            "to": Object {
              "did": "0x6507cde10425d03a8964c5937ccb5760a95d4fd1cda5fc8cdadad8b51e13a38f",
              "number": "0",
            },
          },
        ],
        "result": "Executed",
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5FA8HQPMNmidk6X2h9dfZUEivwRt6ju4cgtpHPEcbB62veMW",
          "5EXPTATZUFcAR9ao5qp9UVaYPoprDpQZFhuVpyibFhA3gmCY",
          "5Hmn2MX8PEd7rBfmXywhAfMvr77FkVy94PyMGncWabDbFVJ7",
          "5DUtBYagbSiwaVdgK54ka26dtdykVn5M71DduDdw64M659EW",
        ],
        "blockId": 294,
        "legs": Array [
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x2795441b98ef5a844c7675df8f6030126a297b898ec2d6029dcf81c72d59e553",
              "number": "0",
            },
            "ticker": "11BTICKER2",
            "to": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
          },
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "11BTICKER1",
            "to": Object {
              "did": "0x2795441b98ef5a844c7675df8f6030126a297b898ec2d6029dcf81c72d59e553",
              "number": "0",
            },
          },
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "11BTICKER1",
            "to": Object {
              "did": "0x553aa64ffcf49987f124633a835e6363a413cfbd10fb2b77ad8bf17afc7606ff",
              "number": "0",
            },
          },
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "11BTICKER1",
            "to": Object {
              "did": "0x2fbb6ef977181d14c7714e134cd5382d42e4214aaf81a2e60b57663946b644c5",
              "number": "0",
            },
          },
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "11BTICKER1",
            "to": Object {
              "did": "0x1c18365371d85ae8662de63bfb0fb90da99b93482e8f043de61ffc83994f8858",
              "number": "0",
            },
          },
        ],
        "result": "Executed",
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5FcXLkN8fkcr65qg8ZJpjZ4Aeqk8HBvXKdDmSK2zUwB4EJsc",
        ],
        "blockId": 327,
        "legs": Array [
          Object {
            "amount": 1000,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "12TICKER",
            "to": Object {
              "did": "0x0854afeca045161ef42ba9be94c973f65efc0e8532933500865a1fd655148f6c",
              "number": "0",
            },
          },
        ],
        "result": "Executed",
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5Ev8nm5CBFm3FRvjso95EbCmVM44WRcZRyH4jgXv6CpLvxTa",
        ],
        "blockId": 354,
        "legs": Array [
          Object {
            "amount": 100000000,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "13TICKER",
            "to": Object {
              "did": "0x6c15581b6ded46773fbb9786b44624fb2d44397f89bfd71b077462c22ac4b72d",
              "number": "0",
            },
          },
        ],
        "result": "Executed",
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5CaWrWYw133scCHRLGBkQv9ssdFFnanPSZhQHmo8ARMeK79Z",
          "5GqANhefumSBPko8oLdKAH8No6Wk7iN4kzgesvqxyQMupfSu",
        ],
        "blockId": 380,
        "legs": Array [
          Object {
            "amount": 1000,
            "from": Object {
              "did": "0xad05ae379784a960d841080123d88b7431bf957a596f0afcea1851902a5f7944",
              "number": "0",
            },
            "ticker": "14TICKERRAIS",
            "to": Object {
              "did": "0x43b3e9575f1f5830fe5451bf1caa1fdee5631ef1644ece1d788250ccb7a95144",
              "number": "0",
            },
          },
        ],
        "result": "Executed",
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5CaWrWYw133scCHRLGBkQv9ssdFFnanPSZhQHmo8ARMeK79Z",
        ],
        "blockId": 383,
        "legs": Array [
          Object {
            "amount": 40,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "14TICKER",
            "to": Object {
              "did": "0xad05ae379784a960d841080123d88b7431bf957a596f0afcea1851902a5f7944",
              "number": "0",
            },
          },
          Object {
            "amount": 0,
            "from": Object {
              "did": "0xad05ae379784a960d841080123d88b7431bf957a596f0afcea1851902a5f7944",
              "number": "0",
            },
            "ticker": "14TICKERRAIS",
            "to": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
          },
        ],
        "result": "Executed",
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GqANhefumSBPko8oLdKAH8No6Wk7iN4kzgesvqxyQMupfSu",
        ],
        "blockId": 384,
        "legs": Array [
          Object {
            "amount": 10,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "14TICKER",
            "to": Object {
              "did": "0x43b3e9575f1f5830fe5451bf1caa1fdee5631ef1644ece1d788250ccb7a95144",
              "number": "0",
            },
          },
          Object {
            "amount": 0,
            "from": Object {
              "did": "0x43b3e9575f1f5830fe5451bf1caa1fdee5631ef1644ece1d788250ccb7a95144",
              "number": "0",
            },
            "ticker": "14TICKERRAIS",
            "to": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
          },
        ],
        "result": "Executed",
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5D9n9c1fCKa6uLf2aZHHFKk3oSPwuPTGvfnjgRbzptzownBH",
        ],
        "blockId": 729,
        "legs": Array [
          Object {
            "amount": "10",
            "from": Object {
              "did": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
              "number": "0",
            },
            "ticker": "15TICKER",
            "to": Object {
              "did": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
              "number": "1",
            },
          },
        ],
        "result": "Executed",
      },
    ],
    "totalCount": 9,
  },
}
`;

exports[`getSettlements test getting settlements with from filter 1`] = `
Object {
  "settlements": Object {
    "__typename": "SettlementsConnection",
    "nodes": Array [
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5DcLNmYXCC8J1hLHJsbtPFNX7kSDx4YTbgGkNYcHUKdvbgU9",
        ],
        "blockId": 262,
        "id": "262/1",
        "legs": Array [
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "11ATICKER",
            "to": Object {
              "did": "0x6507cde10425d03a8964c5937ccb5760a95d4fd1cda5fc8cdadad8b51e13a38f",
              "number": "0",
            },
          },
        ],
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5FA8HQPMNmidk6X2h9dfZUEivwRt6ju4cgtpHPEcbB62veMW",
          "5EXPTATZUFcAR9ao5qp9UVaYPoprDpQZFhuVpyibFhA3gmCY",
          "5Hmn2MX8PEd7rBfmXywhAfMvr77FkVy94PyMGncWabDbFVJ7",
          "5DUtBYagbSiwaVdgK54ka26dtdykVn5M71DduDdw64M659EW",
        ],
        "blockId": 294,
        "id": "294/5",
        "legs": Array [
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x2795441b98ef5a844c7675df8f6030126a297b898ec2d6029dcf81c72d59e553",
              "number": "0",
            },
            "ticker": "11BTICKER2",
            "to": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
          },
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "11BTICKER1",
            "to": Object {
              "did": "0x2795441b98ef5a844c7675df8f6030126a297b898ec2d6029dcf81c72d59e553",
              "number": "0",
            },
          },
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "11BTICKER1",
            "to": Object {
              "did": "0x553aa64ffcf49987f124633a835e6363a413cfbd10fb2b77ad8bf17afc7606ff",
              "number": "0",
            },
          },
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "11BTICKER1",
            "to": Object {
              "did": "0x2fbb6ef977181d14c7714e134cd5382d42e4214aaf81a2e60b57663946b644c5",
              "number": "0",
            },
          },
          Object {
            "amount": 100,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "11BTICKER1",
            "to": Object {
              "did": "0x1c18365371d85ae8662de63bfb0fb90da99b93482e8f043de61ffc83994f8858",
              "number": "0",
            },
          },
        ],
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5FcXLkN8fkcr65qg8ZJpjZ4Aeqk8HBvXKdDmSK2zUwB4EJsc",
        ],
        "blockId": 327,
        "id": "327/1",
        "legs": Array [
          Object {
            "amount": 1000,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "12TICKER",
            "to": Object {
              "did": "0x0854afeca045161ef42ba9be94c973f65efc0e8532933500865a1fd655148f6c",
              "number": "0",
            },
          },
        ],
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
          "5Ev8nm5CBFm3FRvjso95EbCmVM44WRcZRyH4jgXv6CpLvxTa",
        ],
        "blockId": 354,
        "id": "354/1",
        "legs": Array [
          Object {
            "amount": 100000000,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "13TICKER",
            "to": Object {
              "did": "0x6c15581b6ded46773fbb9786b44624fb2d44397f89bfd71b077462c22ac4b72d",
              "number": "0",
            },
          },
        ],
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5CaWrWYw133scCHRLGBkQv9ssdFFnanPSZhQHmo8ARMeK79Z",
        ],
        "blockId": 383,
        "id": "383/6",
        "legs": Array [
          Object {
            "amount": 40,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "14TICKER",
            "to": Object {
              "did": "0xad05ae379784a960d841080123d88b7431bf957a596f0afcea1851902a5f7944",
              "number": "0",
            },
          },
          Object {
            "amount": 0,
            "from": Object {
              "did": "0xad05ae379784a960d841080123d88b7431bf957a596f0afcea1851902a5f7944",
              "number": "0",
            },
            "ticker": "14TICKERRAIS",
            "to": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
          },
        ],
      },
      Object {
        "__typename": "Settlement",
        "addresses": Array [
          "5GqANhefumSBPko8oLdKAH8No6Wk7iN4kzgesvqxyQMupfSu",
        ],
        "blockId": 384,
        "id": "384/6",
        "legs": Array [
          Object {
            "amount": 10,
            "from": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
            "ticker": "14TICKER",
            "to": Object {
              "did": "0x43b3e9575f1f5830fe5451bf1caa1fdee5631ef1644ece1d788250ccb7a95144",
              "number": "0",
            },
          },
          Object {
            "amount": 0,
            "from": Object {
              "did": "0x43b3e9575f1f5830fe5451bf1caa1fdee5631ef1644ece1d788250ccb7a95144",
              "number": "0",
            },
            "ticker": "14TICKERRAIS",
            "to": Object {
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
              "number": "0",
            },
          },
        ],
      },
    ],
    "totalCount": 6,
  },
}
`;
